### DISCUSSION TITLE
Broken benchmark scripts?

### DISCUSSION THREAD

**Discussion by Harris-Miller** [ramda/ramda#3424] [General]
A few recent MRs have touched on the performance for some Ramda functions
https://github.com/ramda/ramda/pull/3417
https://github.com/ramda/ramda/pull/3370

I ran `npm run bench` on my local, a 2023 MacbookPro, M2 Max, 64Gig. Running on Node 18.18.2

<details>
  <summary>Full Results</summary>

  | concat | Hertz | Margin of Error |
  | --- | --- | --- |
  | concat(s1, s2) | 14,718,562 | 2.42% |
  | concat(s1)(s2) | 6,902,866 | 3.44% |
  | concatS1(s2) | 8,157,931 | 0.33% |
  | native concat | 14,871,931 | 0.43% |
  
  | curry | Hertz | Margin of Error |
  | --- | --- | --- |
  | mult4(100) | 38,544,999 | 14.79% |
  | manual(100) | 246,147,972 | 3.72% |
 
  
  | forEach | Hertz | Margin of Error |
  | --- | --- | --- |
  | forEach(x2, nums) | 25,019,194 | 0.21% |
  | forEach(x2)(nums) | 5,080,193 | 0.16% |
  | native forEach | 25,613,509 | 0.75% |
 
  | either | Hertz | Margin of Error |
  | --- | --- | --- |
  | either(gt10, even)(10… | 82,422,925 | 0.31% |
  | either(gt10)(even)(10… | 10,815,278 | 5.51% |
  | either(gt10, even)(8) | 70,802,691 | 4.65% |
  | either(gt10)(even)(8) | 11,111,096 | 4.83% |
  | native gt10(101) || e… | 230,340,544 | 5.74% |
  | native gt10(8) || eve… | 250,120,566 | 1.09% |
 
  | filter where | Hertz | Margin of Error |
  | --- | --- | --- |
  | filter(where({x: isEm… | 121,576 | 1.14% |
  | filter(where({x: isEm… | 122,109 | 0.16% |
  | filterEmptyX(objs) | 121,027 | 1.83% |
  | filter(whereEq({y: fa… | 232,201 | 2.18% |
  | filter(whereEq({y: fa… | 227,156 | 0.80% |
  | filterFalseY(objs) | 236,046 | 0.64% |
  
  | filter | Hertz | Margin of Error |
  | --- | --- | --- |
  | filter(isEven, nums) | 8,145,962 | 2.16% |
  | filter(isEven)(nums) | 5,068,465 | 1.70% |
  | filterEven(nums) | 5,529,485 | 0.21% |
  | native filter | 35,600,798 | 1.64% |
 
  | find where | Hertz | Margin of Error |
  | --- | --- | --- |
  | find(where({x: isEmpt… | 404,731 | 0.51% |
  | find(where({x: isEmpt… | 389,726 | 0.14% |
  | findEmptyX(objs) | 392,070 | 1.72% |
  | find(whereEq({y: fals… | 760,532 | 1.07% |
  | find(whereEq({y: fals… | 698,929 | 1.93% |
  | findFalseY(objs) | 779,224 | 0.68% |
  
  | find | Hertz | Margin of Error |
  | --- | --- | --- |
  | find(isZero, nums) | 16,389,580 | 0.15% |
  | find(isZero)(nums) | 7,267,711 | 2.53% |
  | findZero(nums) | 8,303,808 | 0.46% |
  | native find | 134,473,355 | 0.20% |
 
  
  | findIndex where | Hertz | Margin of Error |
  | --- | --- | --- |
  | findIndex(where({x: i… | 402,133 | 0.59% |
  | findIndex(where({x: i… | 376,699 | 3.37% |
  | findIndexEmptyX(objs) | 389,011 | 0.98% |
  | findIndex(whereEq({y:… | 771,657 | 0.18% |
  | findIndex(whereEq({y:… | 714,946 | 0.46% |
  | findIndexFalseY(objs) | 768,440 | 0.23% |
  
  | findIndex | Hertz | Margin of Error |
  | --- | --- | --- |
  | findIndex(isZero, num… | 16,072,896 | 0.45% |
  | findIndex(isZero)(num… | 7,319,421 | 1.07% |
  | findIndexZero(nums) | 8,286,012 | 0.17% |
  | native findIndex | 135,672,363 | 0.03% |
 
 
  | includes | Hertz | Margin of Error |
  | --- | --- | --- |
  | includes(c, cs) | 17,003,331 | 2.91% |
  | native includes | 20,769,476 | 0.52% |
 
  
  | indexOf | Hertz | Margin of Error |
  | --- | --- | --- |
  | indexOf(sq, nums) | 84,519,692 | 0.19% |
  | indexOf(sq)(nums) | 12,475,150 | 0.71% |
  | idxOf23(nums) | 14,930,324 | 0.12% |
  | native indexOf | 239,766,129 | 0.14% |
 
  | map | Hertz | Margin of Error |
  | --- | --- | --- |
  | map(sq, nums) | 26,919,650 | 0.18% |
  | map(sq)(nums) | 9,135,362 | 1.51% |
  | mapSq(nums) | 10,564,907 | 0.90% |
  | native map | 38,457,477 | 0.16% |
 
  | maxBy | Hertz | Margin of Error |
  | --- | --- | --- |
  | maxBy(computer, nums) | 52,864,144 | 0.15% |
  | maxBy(computer)(vals) | 34,113,558 | 0.44% |
  | maxVal(vals) | 57,074,240 | 0.20% |
  
  | propEq | Hertz | Margin of Error |
  | --- | --- | --- |
  | propEq([1, 2, 3], "va… | 1,315,439 | 0.15% |
  
  | reduce | Hertz | Margin of Error |
  | --- | --- | --- |
  | reduce(add, 0, nums) | 9,547,934 | 0.12% |
  | reduce(add, 0)(nums) | 6,801,443 | 0.11% |
  | reduceAdd(nums) | 7,492,478 | 1.02% |
  | native reduce | 94,406,869 | 0.21% |
 
  | uniq | Hertz | Margin of Error |
  | --- | --- | --- |
  | uniqBy(id, arr1) | 322 | 0.89% |
  | uniqWith(equals, arr1) | 99.58 | 0.69% |
  | uniq(arr1) | 231 | 1.26% |
  | uniqBy(id, arr2) | 36,386 | 0.16% |
  | uniqWith(equals, arr2) | 44.83 | 0.83% |
  | uniq(arr2) | 12,013 | 0.26% |
  | uniqBy(id, arr3) | 1,445 | 0.10% |
  | uniqWith(equals, arr3) | 1,494 | 0.19% |
  | uniq(arr3) | 1,315 | 0.10% |
</details>

Slowdowns when calling functions curried (eg `either(gt10)(even)(8)` versus `either(gt10, even, 8)`) is expected, I was generally surprised at the stark difference between the native functions and ramda's. I decided to narrow down the benchmark.js script so it just ran one test, `find.bench.js`

Group run

| find | Hertz | Margin of Error |
| --- | --- | --- |
| find(isZero, nums) | 16,389,580 | 0.15% |
| find(isZero)(nums) | 7,267,711 | 2.53% |
| findZero(nums) | 8,303,808 | 0.46% |
| native find | 134,473,355 | 0.20% |

Solo run

| find | Hertz | Margin of Error |
| --- | --- | --- |
| find(isZero, nums) | 153,623,249 | 0.22% |
| find(isZero)(nums) | 12,529,704 | 2.19% |
| findZero(nums) | 15,166,628 | 4.00% |
| native find | 134,937,217 | 0.10% |

I'm not sure why, but while the native find comes out to be about the same between the 2 runs, the ramda functions are massively different!

`find(isZero, nums)` is faster than `nums.find(isZero)` here, not surprising given the imperative backing code using a `while` loop

Throwing the same code into jsperf but simplified to the core function yields even a greater magnitude of speed difference

https://jsperf.app/nadupe

<img width="1005" alt="Screenshot 2023-11-15 at 10 47 57 PM" src="https://github.com/ramda/ramda/assets/2763179/157a321d-852a-448b-be38-9d7d1d969a7b">

Is it worth it to try and fix these?

**Comment by kedashoe** [comment]
Looks like the benchmarking library we are using is 11 years old? Probably due for an update..
